drivers/gpio/gpio_rpi_pico.c

```diff
@@ -0,0 +1,191 @@
+
+static int gpio_rpi_configure(const struct device *dev,
+				gpio_pin_t pin,
+				gpio_flags_t flags)
+{
+	if (flags & GPIO_SINGLE_ENDED) {
+		return -ENOTSUP;
+	}
+
+	gpio_init(pin);
+
+	if (flags & GPIO_OUTPUT) {
+		gpio_set_dir(pin, GPIO_OUT);
+
+		if (flags & GPIO_OUTPUT_INIT_HIGH) {
+			gpio_put(pin, 1);
+		} else if (flags & GPIO_OUTPUT_INIT_LOW) {
+			gpio_put(pin, 0);
+		}
+	} else if (flags & GPIO_INPUT) {
+		gpio_set_dir(pin, GPIO_IN);
+		if (flags & GPIO_PULL_UP) {
+			gpio_pull_up(pin);
+		} else if (flags & GPIO_PULL_DOWN) {
+			gpio_pull_down(pin);
+		}
+	}
+
+	return 0;
+}
+
+static int gpio_rpi_port_get_raw(const struct device *dev, uint32_t *value)
+{
+	*value = gpio_get_all();
+	return 0;
+}
+
+static int gpio_rpi_port_set_masked_raw(const struct device *port,
+					uint32_t mask, uint32_t value)
+{
+	gpio_put_masked(mask, value);
+	return 0;
+}
+
+static int gpio_rpi_port_set_bits_raw(const struct device *port,
+					uint32_t pins)
+{
+	gpio_set_mask(pins);
+	return 0;
+}
+
+static int gpio_rpi_port_clear_bits_raw(const struct device *port,
+					uint32_t pins)
+{
+	gpio_clr_mask(pins);
+	return 0;
+}
+
+static int gpio_rpi_port_toggle_bits(const struct device *port,
+					uint32_t pins)
+{
+	gpio_xor_mask(pins);
+	return 0;
+}
+
+static int gpio_rpi_pin_interrupt_configure(const struct device *dev,
+						gpio_pin_t pin,
+						enum gpio_int_mode mode,
+						enum gpio_int_trig trig)
+{
+	struct gpio_rpi_data *data = dev->data;
+	uint32_t events = 0;
+
+	if (mode != GPIO_INT_DISABLE) {
+		if (mode & GPIO_INT_EDGE) {
+			if (trig & GPIO_INT_LOW_0) {
+				events |= GPIO_IRQ_EDGE_FALL;
+			}
+			if (trig & GPIO_INT_HIGH_1) {
+				events |= GPIO_IRQ_EDGE_RISE;
+			}
+		} else {
+			if (trig & GPIO_INT_LOW_0) {
+				events |= GPIO_IRQ_LEVEL_LOW;
+			}
+			if (trig & GPIO_INT_HIGH_1) {
+				events |= GPIO_IRQ_LEVEL_HIGH;
+			}
+		}
+		gpio_set_irq_enabled(pin, events, true);
+	}
+	WRITE_BIT(data->int_enabled_mask, pin, mode != GPIO_INT_DISABLE);
+	return 0;
+}
+
+static int gpio_rpi_manage_callback(const struct device *dev,
+				struct gpio_callback *callback, bool set)
+{
+	struct gpio_rpi_data *data = dev->data;
+
+	return gpio_manage_callback(&data->callbacks, callback, set);
+}
+
+static const struct gpio_driver_api gpio_rpi_driver_api = {
+	.pin_configure = gpio_rpi_configure,
+	.port_get_raw = gpio_rpi_port_get_raw,
+	.port_set_masked_raw = gpio_rpi_port_set_masked_raw,
+	.port_set_bits_raw = gpio_rpi_port_set_bits_raw,
+	.port_clear_bits_raw = gpio_rpi_port_clear_bits_raw,
+	.port_toggle_bits = gpio_rpi_port_toggle_bits,
+	.pin_interrupt_configure = gpio_rpi_pin_interrupt_configure,
+	.manage_callback = gpio_rpi_manage_callback,
+};
+
+static void gpio_rpi_isr(const struct device *dev)
+{
+	struct gpio_rpi_data *data = dev->data;
+	io_irq_ctrl_hw_t *irq_ctrl_base;
+	const io_rw_32 *status_reg;
+	uint32_t events;
+	uint32_t pin;
+
+	irq_ctrl_base = &iobank0_hw->proc0_irq_ctrl;
+	for (pin = 0; pin < NUM_BANK0_GPIOS; pin++) {
```

Comment:
For the future, the pico internal rom has a function to return the address of the righmost '1' bit (in a few cycles), which could be used to set 'pin' rather than increment it. Would this be faster than increment & test each pin blindly?

Reply:
Maybe, but I wouldn't tinker with the ROM for a loop that runs ~30 iterations once at startup

Reply:
the rom-based functions are just as fast (I think) as flash and made visible via the sdk, so no faffing.

I thought this function was being called once for each interrupt, to determine which associated function(s) to call? No?

Reply:
Sorry, thought that was the pinctrl file. I might look into it in the future. I don't think using the ROM is that easy (the ROM should be added to the linker script, etc.)

Reply:
why would it need to be in the linker? Is there some need to record the complete footprint of the program unrelated to loading the code into flash? From the user-of-sdk point of view it's just another sdk call (though in fact it's a macro).